### PR TITLE
Initial commit for the k6 Loki extension

### PR DESCRIPTION
**Pull request by chaudum**
The extension contains functionality to create batches of streams and to
push them to a Loki endpoint using JSON or Protobuf encoding.

This is a first implementation which does not allow to _parametrize all the things_ yet. However, it should be good enough for a start.

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	EncodeJSON bool
+}
+
+func (c *Client) Push(ctx context.Context) (httpext.Response, error) {
+	state := lib.GetState(ctx)
+	if state == nil {
+		return *httpext.NewResponse(ctx), errors.New("state is nil")
+	}
+
+	entries := generateEntries(c.cfg.TenantID, 100)
```
Comment (cyriltovena) [inline client.go:43]: I think we will want to be able to configure more:
- Cardinality generated.
- Format json/apache
- batch size in bytes.

File: batch.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+// newLabelPool creates a "pool" of values for each label name
+// The amount of different values per label name is not configurable yet.
+func newLabelPool(faker *fake.Faker) LabelPool {
+	return LabelPool{
+		Format:    model.LabelValues{"apache_common", "json"}, // needs to match the available flog formats
+		Namespace: generateValues(faker.BS, 10),
+		App:       generateValues(faker.BS, 5),
+		Pod:       generateValues(faker.BS, 100),
```
Comment (chaudum) [inline batch.go:195]: Ideally, there would be JSON definition of these labels so they are not static.

File: batch.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+// sizeBytesAfter returns the size of the batch after the input entry
+// will be added to the batch itself
+func (b *Batch) sizeBytesAfter(entry Entry) int {
+	return b.Bytes + len(entry.Line)
+}
+
+// age of the batch since its creation
+func (b *Batch) age() time.Duration {
+	return time.Since(b.CreatedAt)
+}
```
Comment (chaudum) [inline batch.go:89]: These function are unused at the moment.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+PWD := $(shell pwd)
+GO_FILES := $(shell find . -type f -name "*.go" -print)
+
+k6: $(GO_FILES)
+	xk6 build \
+		--replace "github.com/mingrammer/flog=github.com/chaudum/flog@v0.4.4-0.20211115125504-92153be038e6" \
```
Comment (chaudum) [inline Makefile:6]: `github.com/mingrammer/flog` is not importable, therefore the replacement with [`chaudum/flog`](https://github.com/chaudum/flog/commit/92153be038e65a9383ed50fdf9df1eac8a42fdc0).

Comment (chaudum) [reply Makefile:6]: I will try to get the changes into the upstream project.

File: batch.go (hunk 2 above)
Comment (MStoykov) [reply batch.go:195]: You can take a look at https://github.com/grafana/xk6-client-prometheus-remote/blob/a5db579b304270ae80ca0f689fe35b77c2156041/remote_write.go#L342-L348 if that is what you mean?

**Review by MStoykov — APPROVED** [review]
LGTM from the perspective of should it be fine as an extension. 
I have left some comments but nothing of this is blocking IMO,

File: client.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	url, _ := httpext.NewURL(c.cfg.URL.String(), "push")
+	response, err := httpext.MakeRequest(ctx, &httpext.ParsedHTTPRequest{
+		URL:              &url,
+		Req:              r,
+		Body:             bytes.NewBuffer(buf),
+		Throw:            state.Options.Throw.Bool,
+		Redirects:        state.Options.MaxRedirects,
+		Timeout:          c.cfg.Timeout,
+		ResponseCallback: ResponseCallback,
+	})
```
[the remark this answers is not in the document]
Comment (chaudum) [reply client.go:106]: Yes, I took the implementation from the remote write extension :+1:

File: Makefile (hunk 4 above)
Comment (cyriltovena) [reply Makefile:6]: We can fork this in the repo for sure.

File: batch.go (hunk 3 above)
Comment (cyriltovena) [reply batch.go:89]: I don't think we'll need it.

File: batch.go (hunk 2 above)
Comment (cyriltovena) [reply batch.go:195]: Yeah now that I think about if we generated those randomly, querying will be hard. So we might want to go for namespace1,namespace2.....

Or go with something like cortex.

**Review by cyriltovena — APPROVED** [review]
LGTM

Looks awesome for a firt commit. I think we can polish the data generation next.

File: batch.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		for i := 0; i < n; i++ {
+			entries[i] = Entry{
+				Entry: logproto.Entry{
+					Timestamp: time.Now(),
+					Line:      flog.NewLog(string(labels[model.LabelName("format")]), time.Now()),
+				},
+				TenantID: tenantID,
+				Labels:   labels,
+			}
+			time.Sleep(10 * time.Millisecond)
```
Comment (vlad-diachenko) [inline batch.go:157]: I am not sure that we have to sleep during the data generation. I would add a delay before sending the batch. why do we need it here?

**Review by vlad-diachenko — COMMENTED** [review]
LGTM ;)

**Pull request merged**